### DISCUSSION TITLE
how to define and train a model with mult forward?

### DISCUSSION THREAD

**Discussion by Nightbringers** [google/flax#3456] [General]
here is simple case, one model have mult forward:
class model(nn.Module):

    dtype: jnp.dtype = jnp.float32
    
    def setup(self):
        self.encoder1 = nn.Conv(self.dim, (3, 3),use_bias = False,dtype=self.dtype)
        self.encoder2 = nn.Conv(self.dim, (3, 3),use_bias = False,dtype=self.dtype)
        self.encoder3 = nn.Conv(self.dim, (3, 3),use_bias = False,dtype=self.dtype)
    
      
    def e1(self, obs):
    
        s = self.encoder1(obs)
        return s 
  
    def e2(self,obs):
   
        s = self.encoder2(obs)
        return s 

    def e3(self, obs):
   
        s = self.encoder3(obs)
        return s 

here is the problem: it don't  have __call__ it will raise a error.   And  when train the model, TrainState only have one apply_fn

**Comment by lucaslingle** [comment]
Disclaimer: I'm not a Flax developer but I've used it for about a year now. 

**For apply_fn:** 

Per the [docs](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.train_state.TrainState), you don't have to use TrainState's apply_fn, it is just for convenience. You can just set it to None.

In general, when you want to apply one of the encoders as your forward method, you can call [apply](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.apply) on your instantiated Flax module, and pass in the appropriate forward method required, like so:

```python
Model().apply({"params": params}, inputs, method=Model.e1)
```

**For the \_\_call\_\_ error:**

During initialization, I assume you want to obtain the parameters for all encoders wrapped by the Model class, so would be convenient to define the \_\_call\_\_ method to run all of them. When you run [init](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.apply) with any given method (e.g., ```method=Model.e1```), it will only return the variables/parameters required for that method. So running all encoders with \_\_call\_\_ seems like convenient way to initialize all encoders' parameters and solve your error for \_\_call\_\_ as well. 

**Misc.:**
If you plan on using the encoders in different circumstances during training, you might want to just call the encoders directly, as you might not need a wrapper class. However, in this case, you would have to manage multiple TrainState objects, which might be inconvenient. This depends on your codebase and requirements.
